File: src/main/java/cpsc224/creatures/Creature.java
package cpsc224.creatures;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import cpsc224.damagetypes.DamageType;
import cpsc224.effects.Effect;
import cpsc224.items.Inventory;

import javax.swing.*;

/**
 * A class that represents creatures within the game.
 */

public class Creature {
    private double health;
    private double maxHealth;
    private String name;
    private CreatureModifiers baseModifiers;
    private CreatureModifiers turnModifiers;
    private List<Effect> effects = new ArrayList<>();
    private Inventory inventory;
    private ImageIcon icon;
    private List<Double> weaponWeights;

    /**
     * Instantiates a creature.
     * @param creatureName the name
     * @param maxHealth the max health
     * @param baseModifiers the base modifiers
     * @param inventory the inventory
     */
    public Creature(String creatureName, double maxHealth, CreatureModifiers baseModifiers, Inventory inventory, List<Double> weaponWeights) {
        this(creatureName, maxHealth, baseModifiers, inventory, null, weaponWeights);
    }

    /**
     * Instantiates a creature.
     * @param creatureName the name
     * @param maxHealth the max health
     * @param baseModifiers the base modifiers
     * @param inventory the inventory
     * @param icon the ImageIcon
     */
    public Creature(String creatureName, double maxHealth, CreatureModifiers baseModifiers, Inventory inventory, ImageIcon icon) {
        this.name = creatureName;
        this.maxHealth = maxHealth;
        this.health = maxHealth;
        this.baseModifiers = baseModifiers;
        this.inventory = inventory;
        this.icon = icon;
        weaponWeights = new ArrayList<>();

        resetTurnModifiers();
    }

    /**
     * Instantiates a creature.
     * @param creatureName the name
     * @param maxHealth the max health
     * @param baseModifiers the base modifiers
     * @param inventory the inventory
     * @param icon the ImageIcon
     */
    public Creature(String creatureName, double maxHealth, CreatureModifiers baseModifiers, Inventory inventory, ImageIcon icon, List<Double> weaponWeights) {
        this.name = creatureName;
        this.maxHealth = maxHealth;
        this.health = maxHealth;
        this.baseModifiers = baseModifiers;
        this.inventory = inventory;
        this.icon = icon;
        this.weaponWeights = weaponWeights;

        resetTurnModifiers();
    }

    /**
     * Adds an effect to the creature. The effect is
     * only applied if it is an instantly applied effect.
     * @param effect the effect
     * @return the result of the effect as a string
     */
    public String addEffect(Effect effect) {
        String result = null;

        // apply effect if it is applied instantly and set result accordingly
        if (effect.isAppliedInstantly()) {
            result = effect.applyEffect(this);
        } else {
            result = "Added " + effect.toString() + " to " + name;
        }

        // don't add the effect if it is out of turns
        if (effect.getTurns() <= 0)
            return result;

        effects.add(effect);
        return result;
    }

    /**
     * Clears all effects from the creature.
     */
    public void clearEffects() {
        effects.clear();
    }

    /**
     * Gets all effects from the creature.
     * @return the effects as a collection
     */
    public Collection<Effect> getEffects() {
        return effects;
    }

    /**
     * Applies all effects to the creature for the turn. 
     * The creature's turn modifiers are reset to the base modifiers.
     * Effects that run out of turns are removed.
     * @return the result as a String
     */
    public String calculateEffects() {
        resetTurnModifiers();

        List<String> results = new ArrayList<>();
        for (int i = 0; i < effects.size(); i++) {
            Effect e = effects.get(i);
            results.add(e.applyEffect(this));

            if (e.getTurns() <= 0) {
                effects.remove(i);
                i--;
            }
        }
        return String.join(", ", results);
    }

    /**
     * Gets the name of the creature.
     * @return the name
     */
    public String getName(){
        return name;
    }

    /**
     * Gets the current health of the creature.
     * @return the health
     */
    public double getHealth() {
        return health;
    }

    /**
     * Sets the current health of the creature.
     * @param health the new health
     */
    public void setHealth(double health) {
        this.health = health;
    }

    /**
     * Gets the max health of the creature.
     * @return the max health
     */
    public double getMaxHealth() {
        return maxHealth;
    }

    /**
     * Gets the base modifiers of the creature.
     * @return the base modifiers
     */
    public CreatureModifiers getBaseModifiers() {
        return baseModifiers;
    }

    /**
     * Gets the turn modifiers of the creature.
     * @return the turn modifiers
     */
    public CreatureModifiers getTurnModifiers() {
        return turnModifiers;
    }

    /**
     * Gets the inventory of the creature.
     * @return the inventory
     */
    public Inventory getInventory() {
        return inventory;
    }

    public ImageIcon getIcon() {
        return icon;
    }

    public List<Double> getWeaponWeights(){
        return weaponWeights;
    }

    public void setWeaponWeights(List<Double> weights){
        weaponWeights = weights;
    }

    /**
     * Adds health to the creature up to the max health.
     * @param amount the amount of health to add
     * @return the amount of health added
     */
    public double addHealth(int amount) {
        double amountAdded = Math.min(maxHealth - health, amount);
        health = Math.min(maxHealth, health + amount);
        return amountAdded;
    }

    /**
     * Applies damage to the creature. Takes into account damage type resistances.
     * @param damage the amount of damage
     * @param type the damage type
     * @return the amount of damage applied
     */
    public double applyDamage(double damage, DamageType type) {
        double damageDealt = damage * turnModifiers.getResistance(type);

        if (health - damageDealt < 0)
            health = 0;
        else
            health -= damageDealt;

        return damageDealt;
    }

    /**
     * Applies damage as a percent of the creature's max health.
     * @param percent the percent of the creature's health
     * @param type the damage type
     * @return the amount of damage applied
     */
    public double applyPercentDamage(double percent, DamageType type) {
        return applyDamage(maxHealth * percent, type);
    }

    /**
     * resets the turn modifiers to a copy of the base modifiers
     */
    private void resetTurnModifiers(){
        turnModifiers = baseModifiers.clone();
    }
}

File: web/src/game/creatures/Creature.ts
import type { DamageType } from '../damagetypes/DamageType';
import type { Effect } from '../effects/Effect';
import { Inventory } from '../items/Inventory';
import type { IconRef } from '../utils/icons';
import { CreatureModifiers } from './CreatureModifiers';
import type { Coordinate } from '../levels/Coordinate';

export class Creature {
  protected name: string;
  protected maxHealth: number;
  protected health: number;
  protected baseModifiers: CreatureModifiers;
  protected turnModifiers: CreatureModifiers;
  protected inventory: Inventory;
  protected icon: IconRef;
  protected weaponWeights: number[];
  protected effects: Effect[] = [];
  protected currentPosition: Coordinate | null = null;

  constructor(
    name: string,
    maxHealth: number,
    baseModifiers: CreatureModifiers,
    inventory: Inventory,
    icon: IconRef = null,
    weaponWeights: number[] = [],
  ) {
    this.name = name;
    this.maxHealth = maxHealth;
    this.health = maxHealth;
    this.baseModifiers = baseModifiers;
    this.turnModifiers = baseModifiers.clone();
    this.inventory = inventory;
    this.icon = icon;
    this.weaponWeights = [...weaponWeights];
  }

  getName(): string {
    return this.name;
  }

  setName(name: string): void {
    this.name = name;
  }

  getHealth(): number {
    return this.health;
  }

  setHealth(health: number): void {
    this.health = Math.max(0, Math.min(this.maxHealth, health));
  }

  getMaxHealth(): number {
    return this.maxHealth;
  }

  setMaxHealth(maxHealth: number): void {
    this.maxHealth = maxHealth;
  }

  /** Re-clone turn mods after mutating base resists/stats. */
  syncTurnModifiersFromBase(): void {
    this.turnModifiers = this.baseModifiers.clone();
  }

  getBaseModifiers(): CreatureModifiers {
    return this.baseModifiers;
  }

  getTurnModifiers(): CreatureModifiers {
    return this.turnModifiers;
  }

  getInventory(): Inventory {
    return this.inventory;
  }

  getIcon(): IconRef {
    return this.icon;
  }

  getWeaponWeights(): number[] {
    return this.weaponWeights;
  }

  getEffects(): Effect[] {
    return this.effects;
  }

  getCurrentPosition(): Coordinate | null {
    return this.currentPosition;
  }

  setCurrentPosition(pos: Coordinate): void {
    this.currentPosition = pos;
  }

  addEffect(e: Effect): string {
    let message = '';
    if (e.isAppliedInstantly) {
      message = e.applyEffect(this);
    } else {
      message = `Added ${e.toString()} to ${this.name}`;
    }
    if (e.getTurns() <= 0) {
      return message;
    }
    this.effects.push(e);
    return message;
  }

  clearEffects(): void {
    this.effects = [];
    this.turnModifiers = this.baseModifiers.clone();
  }

  removeEffectsByName(name: string): number {
    const before = this.effects.length;
    this.effects = this.effects.filter((e) => e.getName() !== name);
    return before - this.effects.length;
  }

  calculateEffects(): string {
    this.turnModifiers = this.baseModifiers.clone();
    const messages: string[] = [];
    const remaining: Effect[] = [];
    for (const e of this.effects) {
      messages.push(e.applyEffect(this));
      if (e.getTurns() > 0) remaining.push(e);
    }
    this.effects = remaining;
    return messages.filter(Boolean).join(', ');
  }

  addHealth(amount: number): number {
    const before = this.health;
    this.health = Math.min(this.maxHealth, this.health + amount);
    return this.health - before;
  }

  applyDamage(dmg: number, type: DamageType): number {
    const dealt = dmg * this.turnModifiers.getResistance(type);
    this.health = Math.max(0, this.health - dealt);
    return dealt;
  }

  applyPercentDamage(pct: number, type: DamageType): number {
    return this.applyDamage(this.maxHealth * pct, type);
  }

  isAlive(): boolean {
    return this.health > 0;
  }
}
